bio2zarr/ts.py

```diff
@@ -0,0 +1,271 @@
+        individuals = np.unique(ts.tables.nodes.individual[ts.samples()])
+        if len(individuals) == 1 and individuals[0] == tskit.NULL:
+            # No samples refer to individuals
+            individuals = None
+        else:
+            # np.unique sorts the argument, so if NULL (-1) is present it
+            # will be the first value.
+            if individuals[0] == tskit.NULL:
+                raise ValueError(
+                    "Sample nodes must either all be associated with individuals "
+                    "or not associated with any individuals"
+                )
+
+        if individuals is not None:
+            self.sample_ids = []
+            for i in individuals:
+                if i < 0 or i >= self.ts.num_individuals:
+                    raise ValueError("Invalid individual IDs provided.")
+                ind = self.ts.individual(i)
+                if len(ind.nodes) == 0:
+                    raise ValueError(f"Individual {i} not associated with a node")
+                is_sample = {ts.node(u).is_sample() for u in ind.nodes}
+                if len(is_sample) != 1:
+                    raise ValueError(
+                        f"Individual {ind.id} has nodes that are sample and "
+                        "non-samples"
+                    )
+                self.sample_ids.extend(ind.nodes)
+                self.individual_ploidies.append(len(ind.nodes))
+                self.max_ploidy = max(self.max_ploidy, len(ind.nodes))
+        else:
+            if ploidy is None:
+                ploidy = 1
+            if ploidy < 1:
+                raise ValueError("Ploidy must be >= 1")
+            if ts.num_samples % ploidy != 0:
+                raise ValueError("Sample size must be divisible by ploidy")
+            self.individual_ploidies = np.full(
+                ts.num_samples // ploidy, ploidy, dtype=np.int32
+            )
+            self.max_ploidy = ploidy
+            self.sample_ids = np.arange(ts.num_samples, dtype=np.int32)
+
+        self.num_samples = len(self.individual_ploidies)
+
+        self.samples = [vcz.Sample(id=f"tsk_{j}") for j in range(self.num_samples)]
+
+    def iter_alleles(self, start, stop, num_alleles):
+        for variant in self.ts.variants(
+            samples=self.sample_ids,
+            isolated_as_missing=self.isolated_as_missing,
+            left=self.positions[start],
+            right=self.positions[stop] if stop < self.num_records else None,
+        ):
+            alleles = np.full(num_alleles, constants.STR_FILL, dtype="O")
+            for i, allele in enumerate(variant.alleles):
+                assert i < num_alleles
+                alleles[i] = allele
+            yield alleles
+
+    def iter_contig(self, start, stop):
+        yield from (0 for _ in range(start, stop))
+
+    def iter_field(self, field_name, shape, start, stop):
+        if field_name == "position":
+            for pos in self.ts.tables.sites.position[start:stop]:
```

Comment:
``self.ts.sites_position`` - you're creating a copy of the tables every time here!